wip on scala3

pityka opened the pull request.

pityka commented:
The coreJVM module in this branch compile on 2.13 and 3. the test:coreJVM target compiles on 2.13 but not on 3

pityka commented:
@tnielens , the scala3 PR passes CI. Do you want to take a look? If you are not interested then just tell me. But in either case I don't plan to merge this very soon.

tnielens commented on build.sbt:
```diff
@@ -15,52 +18,65 @@ ThisBuild / libraryDependencySchemes ++= Seq(
 )
 
-ThisBuild / versionPolicyIntention := Compatibility.BinaryCompatible
+ThisBuild / versionPolicyIntention := Compatibility.None
```
No binary compat?

pityka replied:
The main reason is this:

```scala
trait ScalarTag[@spec(Boolean, Int, Long, Float, Double) T]
    extends ClassTag[T] { .. }

type CLM[C] = ClassTag[C]

class ScalarTagAny[T: CLM] extends ScalarTag[T]
```
ScalarTagAny is recursive because itself is a ClassTag. This was working in scala2 because stAny has no type annotation thereby is not eligible for recursion in:

```
implicit def stAny[T: CLM] = new ScalarTagAny[T] { ... }
```
. If we place the type annotation then it blows up in runtime (scala3 needs that). I did not try it with an implicit class but that can't be top level and ScalarTagAny is top level in the current main branch.

I reworked that part by making ScalarTag compose with ClassTag and not inherit it. 
Loosely connected to this I found some parts which I deemed hard to maintain (scalarTagAnyVal vs scalaTagAny vs scalarTagAnyRef and the takeType in Frame) so I removed those.

I did not remove but unless the implicit conversions for `na` can be fixed that is also a candidate for removal. It is very confusing that there is an `NA` and an `na` object.

tnielens replied:
> I did not remove but unless the implicit conversions for na can be fixed that is also a candidate for removal. It is very confusing that there is an NA and an na object.

Agreed. Also the fact that `na` plays poorly with type inference. For example, `Vec(1, na)` doens't infer as `Vec[Int](1, na)`.

> `class ScalarTagAny[T: CLM] extends ScalarTag[T]`

Oho, didn't notice this unnecessary composition and inheritance pattern. Nice catch.

tnielens replied:
Addressed in 58b38da.

tnielens commented on saddle-core/src/main/scala/org/saddle/Frame.scala:
```diff
@@ -532,38 +535,6 @@ class Frame[RX: ST: ORD, CX: ST: ORD, @spec(Int, Long, Double) T](
-    * @tparam U
-    *   The type of columns to extract
-    */
-  def colType[U: ST]: Frame[RX, CX, U] = {
```
were these moved or removed?

tnielens commented on saddle-core/src/main/scala/org/saddle/Series.scala:
```diff
@@ -925,12 +926,12 @@ class Series[X: ST: ORD, @spec(Int, Long, Double) T: ST](
     */
   def proxyWith(
       proxy: Series[X, T]
-  )(implicit fn: org.saddle.scalar.NA.type => T): Series[X, T] = {
```
what's the issue here?

pityka replied:
Indeed, this can be reverted.


The implicit conversion (application of `fn`) does not trigger.

tnielens commented on saddle-core/src/test/scala/org/saddle/SeriesCheck.scala:
```diff
@@ -501,7 +503,7 @@ class SeriesCheck extends Specification with ScalaCheck {
         .reindex(Index(10, 11, 12, 13, 20, 21, 22, 23), FillForward, limit=1)
         .values
-        .must_==(Vec[Int](1, 1, na, na, 2, 2, na, na))
+        .must_==(Vec[Int](1, 1, naI, naI, 2, 2, naI, naI))
```
No implicit conversion?

pityka replied:
implicit conversion did not happen, even if naToInt is explicitly imported.

tnielens replied:
Doesn't scala 3 require a specific import to enable implicit conversions?

tnielens replied:
The implicit conversions still work, but implicit function values aren't used in implicit conversions in scala 3. They must be converted into regular methods.
```scala
//     implicit val naToInt: na.type => Int = (_: na.type) => scalar.ScalarTagInt.missing
implicit def naToInt(@unused it: na.type): Int = scalar.ScalarTagInt.missing
```

tnielens replied:
If backward compat must be preserved, this could be done:
```scala
// remove implicit modifier of the function value
val naToInt: na.type => Int = (_: na.type) => scalar.ScalarTagInt.missing
implicit def naToIntDef(@unused it: na.type): Int = scalar.ScalarTagInt.missing
```

tnielens replied:
Reference https://docs.scala-lang.org/scala3/reference/changed-features/implicit-conversions-spec.html#differences-with-scala-2-implicit-conversions

tnielens replied:
Addressed in 58b38da.

tnielens commented on saddle-core/src/test/scala/org/saddle/VecCheck.scala:
```diff
@@ -363,54 +365,54 @@ class VecCheck extends Specification with ScalaCheck {
       v.fillBackward() mustEqual Vec[Int](2, 2, 2, 1)
     }
 
     "fillBackword fills values backward until the limit if greater than 0" in {
```
unrelated to this PR but typo in Backword

tnielens commented on saddle-ops-inlined/src/test/scala/org/saddle/macros/MatAbritraries.scala:
```diff
@@ -0,0 +1,39 @@
+/** Copyright (c) 2013 Saddle Development Team
```
Were these files copied? Why?

pityka replied:
using "test->test" dependency produced an error.  The problem was that it put both the 2.13 and the 3 artifacts on the class path. Not using "test->test" submodule dependencies cleared it up.

Looked like an unwanted behavior in sbt around the interaction of the "test->test" feature and the crossScalaVersions feature.

tnielens replied:
I'd like to look further into this one, prevent code duplication.

tnielens commented on build.sbt:
```diff
@@ -231,6 +249,8 @@ lazy val stats = project
   .settings(commonSettings: _*)
   .settings(
     name := "saddle-stats",
+        crossScalaVersions := Seq(scalaVersion213),
```
needs formatting

tnielens commented on saddle-core/src/main/scala/org/saddle/util/OrderInstances.scala:
```diff
@@ -384,6 +391,27 @@ trait OrderInstances {
       T16: ORD,
       T17: ORD,
       T18: ORD
+  ]: ORD[
```
you might consider disabling formatting in these cases

pityka commented on saddle-core/src/main/scala/org/saddle/scalar/ScalarTagAny.scala:
```diff
@@ -22,6 +22,8 @@ import org.saddle.Buffer
 
 class ScalarTagAny[T: CLM] extends ScalarTag[T] {
+
+  val classTag = implicitly[CLM[T]]
```
@tnielens this can be removed as well, right?

tnielens replied:
Yes guess so.

tnielens commented:
It might be possible to resolve all mima errors. I'm reverting some legit bug fixes done here but causing binary compat breakage and listing them in #437.

pityka commented:
Thanks for figuring those out!

pityka commented:
I believe I fixed the remaining issues to clean the build. However I rebased on master first, the needed changes are in this branch: https://github.com/pityka/saddle/tree/feature/scala3-2nd  I suggest that you rebase this branch (feature/scala3) then cherry pick and review these:

540a5617886d168b558e1d70956b942c2a743127 (adds back collection compat in the build)
212696466826f799897211a0df43e24cfb9eb8b1 (adds back an internal dependency in the 2.13 build)
8e661cc3b63116e93607cfe6f921b4510d5f1fac (removes a source dependency on the collection-compat, does not touches the build, this source dependency was introduced in the main branch)
15b66ac2b30b4486ae48305c8c0daa36a5d5e5da (scala 3.1.2)
